[FLINK-10002] [Webfrontend] WebUI shows jm/tm logs more friendly.

zhangxinyu1 opened the pull request:
## What is the purpose of the change
This commit makes WebUI show jm/tm logs more friendly. WebUI can show historical logs by inputing a filename and a offset range to read of the file, shown as follows.
![image](https://user-images.githubusercontent.com/12528894/47417697-20b36b00-d768-11e8-8268-35990c08bc22.png)
![image](https://user-images.githubusercontent.com/12528894/47417728-30cb4a80-d768-11e8-82ee-94d39edca151.png)
![image](https://user-images.githubusercontent.com/12528894/47417747-388aef00-d768-11e8-86b9-ae7489742c20.png)
![image](https://user-images.githubusercontent.com/12528894/47417777-480a3800-d768-11e8-8652-c9f47a45c225.png)


This pr adds 6 new REST APIs:

- For JobManager
`/jobmanager/log?filename=xxx&start=0&size=10240`    Return logs of the current log  file or a given file in the range.
`/jobmanager/loglist`    Return the list of historical log filename.
`/jobmanager/stdout?start=0&size=10240`    Return the stdout data in the range

- For TaskManager
`/taskmanagers/:taskmanagerid/log?filename=xxx&start=0&size=10240`    Return logs of the current log file or a given file in the range.
`/taskmanagers/:taskmanagerid/loglist`    Return the list of historical log filename.
`/taskmanagers/:taskmanagerid/stdout?start=0&size=10240`    Return the stdout data in the range

## Brief change log
This pr mainly add 6 APIs in new (WebMonitorEndpoint) REST API. Besides, some related frontend codes are changed, such as

1.  `JobManagerLogFileHandler` is used to handle the request to fetch stdout and log file of job manager. The related headers are `JobManagerLogFileHeaders` and `JobManagerStdoutFileHeaders`.
2.  `JobManagerLogListHandler` is used to handle the request to fetch log list of job manager. The related header is `JobManagerLogListHeaders`, and return a `LogListInfo` value.
3. `TaskManagerLogHandler` is used to handle the request to fetch stdout and log file of a given task manager. The related headers are `TaskManagerLogFileHeaders` and `TaskManagerStdoutFileHeaders`.
4.  `TaskManagerLogListHandler` is used to handle the request to fetch log list of a given task manager. The related header is `TaskManagerLogListHeaders`.
5.  `FileOffsetRange` is designed to identify a range of log files, which includes a start offset and an end offset. `LogFilenameQueryParameter`, `LogSizeQueryParameter` and `LogStartOffsetQueryParameter` are the parameters of requests.

## Verifying this change
  - `JobManagerLogFileHandlerTest.java` is used to test the request “/jobmanager/log”
  - `JobManagerLogListHandlerTest.java` is used to test the request "/jobmanager/loglist". 
  - `TaskManagerLogListHandler.java` is used to test "/taskmanagers/:taskmanagerid/loglist"
  - Method `testRequestUploadFile()` in `TaskExecutor.java` is used to test whether task manager can upload the taskmanager log within a given offset range correctly.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no 
  - Anything that affects deployment or recovery: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

zhangxinyu1 commented:
@yanghua @tillrohrmann Could you please review this pr?

zhangxinyu1 commented:
@buptljy Could you please take a look at the pr.

tillrohrmann commented:
Thanks for your contribution @zhangxinyu1. From a high level perspective, this adds a good feature. At the moment I'm quite busy and it will take some time for me to review your PR. Please bear with us until then.

zhangxinyu1 commented:
@tillrohrmann Thanks, I'll resolve conflicts if it has until you review this PR.

zhangxinyu1 commented:
@yanghua Thanks. Conflits are resovled.

ifndef-SleePy commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskManagerMessages.scala:
```diff
@@ -117,11 +118,13 @@ object TaskManagerMessages {
-  /** Indicates a request for the .log file */
-  case object LogFileRequest extends LogTypeRequest
+  /** Indicates a request for the log file */
+  case class LogFileRequest(filename : String, range: FileOffsetRange) extends LogTypeRequest
```
Just leave the legacy codes alone.

ifndef-SleePy commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -338,6 +340,9 @@ class TaskManager(
             "available. Cannot upload TaskManager logs."))
       }
 
+    case LogListRequest =>
```
Legacy problem again.

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/files/StaticFileServerHandler.java:
```diff
@@ -104,6 +106,8 @@
 	/** Be default, we allow files to be cached for 5 minutes. */
 	private static final int HTTP_CACHE_SECONDS = 300;
 
+	private static final Pattern currentLogPattern = Pattern.compile("^/jobmanager/(log|stdout)(/([^/]*))?$");
```
Oh that's my fault. We still need this in master side.
Ignore my comment upside.

zhangxinyu1 commented:
@ifndef-SleePy Thank you so much for your suggestions. I'll modify the code this week.

zhangxinyu1 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/files/FileOffsetRange.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class FileOffsetRange implements Serializable{
+	private static final String SEPARATOR = "-";
+	private final long start;
```
The RESTful API like "/xxx/taskmanager/$filename?start=123&size=456" is better to use. However, I think 'Range' is always understood as something from 'start' to 'end'.

zhangxinyu1 replied:
The RESTful API like "/xxx/taskmanager/$filename?start=123&size=456" is better to use. However, I think 'Range' is always understood as something from 'start' to 'end'.

zhangxinyu1 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/TaskManagerStdoutFileHandler.java:
```diff
@@ -54,7 +55,7 @@ public TaskManagerStdoutFileHandler(
 	@Override
-	protected CompletableFuture<TransientBlobKey> requestFileUpload(ResourceManagerGateway resourceManagerGateway, ResourceID taskManagerResourceId) {
-		return resourceManagerGateway.requestTaskManagerFileUpload(taskManagerResourceId, FileType.STDOUT, timeout);
+	protected CompletableFuture<TransientBlobKey> requestFileUpload(ResourceManagerGateway resourceManagerGateway, ResourceID taskManagerResourceId, String filename, FileOffsetRange range) {
```
".out" filename wouldn't be in log list.  The log list only includes filename of all historical logs. If we modify the implement of `LogListHandler`, we can also get the stdout filename. However, do you think we should mix them up?

zhangxinyu1 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerLogFileWithRangeHeaders.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Headers for the {@link TaskManagerLogFileHandler}.
+ */
+public class TaskManagerLogFileWithRangeHeaders implements UntypedResponseMessageHeaders<EmptyRequestBody, TaskManagerMessageParameters> {
```
I'm wondering should the current implementation be compatible with previous interfaces `/taskmanagers/:taskmanagerId/log` and `/taskmanagers/:taskmanagerId/stdout`?

zhangxinyu1 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/RangePathParameter.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Path parameter identifying the offset range of a log file to read.
+ */
+public class RangePathParameter extends MessagePathParameter<FileOffsetRange> {
```
Good idea!

zhangxinyu1 replied:
`/taskmanagers/:%s/log?filename=xxx&start=123&size=456` may be better since it is compatible with the old API `/taskmanagers/:%s/log`, in which `filename`, `start` and `size` are all optional.

zhangxinyu1 commented:
@ifndef-SleePy Thanks for your suggestions. I have modified the code. I removed the legacy part and refactor the implementation. 
About the test, I don't find any tests for the REST API of logs. Should I add tests for `JobManagerLogFileHandler`, `JobManagerLogListHandler`, `TaskManagerLogListHandler ` and `TaskManagerLogFileHandler `?

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/FileOffsetRange.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * FileOffsetRange is used to decide which part of files to read.
+ */
+public class FileOffsetRange implements Serializable{
```
We should implement the `hashCode` and `equals` methods. Because the `FileOffsetRange` would be used in cache key. These methods would be used to decide whether the query hit the cache or not. If these methods are missing, the cache would be always missing.

zhangxinyu1 replied:
ok

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -814,14 +819,18 @@ public void heartbeatFromResourceManager(ResourceID resourceID) {
 		switch (fileType) {
 			case LOG:
-				filePath = taskManagerConfiguration.getTaskManagerLogPath();
+				if(filename != null && !filename.equals("")) {
```
```suggestion
				if (filename != null && !filename.equals("")) {
```

ifndef-SleePy commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:
```diff
@@ -834,10 +843,20 @@ public void heartbeatFromResourceManager(ResourceID resourceID) {
+				final RandomAccessFile raf;
+				try {
+				 raf = new RandomAccessFile(file, "r");
+				}catch (FileNotFoundException e) {
```
```suggestion
				} catch (FileNotFoundException e) {
```

ifndef-SleePy commented:
> About the test, I don't find any tests for the REST API of logs. Should I add tests for `JobManagerLogFileHandler`, `JobManagerLogListHandler`, `TaskManagerLogListHandler ` and `TaskManagerLogFileHandler `?

Yes, you should add some unit tests to make sure the basic logic is right. Some old codes may not be covered well by testing. However the newer codes are covered quite well. We should keep this good habit.

zhangxinyu1 commented:
> I think we missed an important thing. We should offer file size to user. If not, the parameters 'start', 'end' or 'size' would be useless since the user may have no idea how to set these parameters.

Maybe we can set the start=0 and size=fileSize as the default value when user choose a log file.

zhangxinyu1 commented:
Three classes: `JobManagerLogFileHandlerTest.java`, `JobManagerLogListHandlerTest.java`,  `TaskManagerLogListHandler.java` and a method `testRequestUploadFile` in TaskExecutor.java are added for unit testing.

zhangxinyu1 commented:
@yanghua @tillrohrmann @ifndef-SleePy Sorry to bother you. Could you please look at this pr if you have time?

zentol commented:
Outdated and likely subsumed by FLINK-13987.
